### ISSUE TITLE
IPolicyForHTTPS.createForNetloc incorrectly documents its hostname parameter's type

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7302]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)| @habnabit reported|
|-|-|
|Trac ID|trac#7302|
|Type|defect|
|Created|2014-05-06 16:06:14Z|
|Branch|https://github.com/twisted/twisted/tree/fix-IPolicyForHTTPS-documentation-7302|

`IPolicyForHTTPS.createForNetloc` states that the `hostname` parameter should be of type `unicode`. However, both of its implementors seem to disagree. `BrowserLikePolicyForHTTPS` documents that its `hostname` parameters is `bytes` and attempts `hostname.decode('ascii')`. `_DeprecatedToCurrentPolicyForHTTPS` similarly documents its `hostname` parameter is `bytes`. The current primary user of the interface, `Agent`, does not explicitly document that it passes bytes, but regardless only ever passes bytes to `createForNetloc`. One test does document the parameter as `unicode`, but then passes bytes anyway.

Based on the above, it's my opinion that the action to be taken is to change the documentation to be consistent with the implementation.

<details><summary>Searchable metadata</summary>

```
trac-id__7302 7302
type__defect defect
reporter__habnabit habnabit
priority__normal normal
milestone__None None
branch__branches_fix_IPolicyForHTTPS_documentation_7302 branches/fix-IPolicyForHTTPS-documentation-7302
branch_author__habnabit habnabit
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1399392374000000 1399392374000000
changetime__1445217216921098 1445217216921098
version__None None
owner__habnabit habnabit
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_2"></a>|@habnabit commented|
|-|-|

(In [42612]) Branching to 'fix-IPolicyForHTTPS-documentation-7302'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_3"></a>|@habnabit commented|
|-|-|

(In [42613]) Update 'unicode' -> 'bytes'.

refs [#7302](https://github.com/twisted/twisted/issues/7302)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_4"></a>|@habnabit commented|
|-|-|

[Forced a build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/fix-IPolicyForHTTPS-documentation-7302).

**Comment by twisted-trac** [comment]
|[<img alt="jerith's avatar" src="https://avatars.githubusercontent.com/u/43731?s=50" width="50" height="50">](https://github.com/jerith)<a name="note_5"></a>|@jerith commented|
|-|-|

Related to this, ```BrowserLikePolicyForHTTPS.createForNetloc``` documents a ```tls``` parameter which isn't present.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

> Related to this, `BrowserLikePolicyForHTTPS.createForNetloc` documents a `tls` parameter which isn't present.

Please file a ticket for it instead of commenting on an already-up-for-review ticket which a different scope.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

This looks good. Thanks for the fix.

Since BrowserLikePolicyForHTTPS implements IPolicyForHTTPS, does it need to duplicate docstring for methods which are already documented in the interface?
Maybe a ```See iweb.BrowserLikePolicyForHTTPS.createForNetloc``` can help with reducing code duplication.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_9):
> This looks good. Thanks for the fix.
> 
> Since BrowserLikePolicyForHTTPS implements IPolicyForHTTPS, does it need to duplicate docstring for methods which are already documented in the interface?
> Maybe a ```See iweb.BrowserLikePolicyForHTTPS.createForNetloc``` can help with reducing code duplication.

Nope.  pydoctor will copy the interface docstring if it implements the interface, so it's actually better to leave it blank.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set owner to @habnabit|
|-|-|

Sigh.  URLs are weird.  I think maybe this was a mistake?  But you're right in that it's a mistake that was already made, and we should fix the documentation to reflect reality.  Fixing the mistake for real might involve IRIs or something.

Make it so.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Ahem.  Ping.  Please land within 24 hours or I will.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set status to `closed`|
|-|-|

(In [42954]) Merge fix-IPolicyForHTTPS-documentation-7302: fix interface type documentation

Author: habnabit

Reviewer: glyph

Fixes: [#7302](https://github.com/twisted/twisted/issues/7302)

IPolicyForHTTPS.createForNetloc states that the hostname parameter should be of
type unicode. However, both of its implementors seem to disagree.  This change
adjusts the interface documentation to reflect the reality of the
implementation.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

Ticket retargeted after milestone deleted
